### PR TITLE
Versioning Change

### PR DESCRIPTION
**Pull request by shaopu1225**
- Integrate Data placement and data transfer policy to prototype. This can be checked under `policy` folder.
- Fix GCS delete marker abnormal behavior (GCS does not have delete marker like AWS). This is done in `skyproxy.rs`.
- Change default DB server to postgres to get better performance under high concurrency scenarios. in `db.py`.
- Support multiple workers in uvicorn, use shared_memory to solve concurrency issues. (May need further testing with logging/fetching in Postgres) The changes are mostly in `db.py` and `app.py`. 
- Add record metrics files & methods. The file used for sending record metrics is `send.py`, the corresponding data structures are defined in `object_schemas`. The recording operations are currently commented, can be found in `skyproxy.rs`.
- Refractor test cases in `test_app.py` and `test_version_app.py` and `skyproxy_test.rs` to test/pass the changes.

### CODE REVIEW

File: s3-proxy/justfile [hunk 1]
```diff
@@ -11,7 +11,7 @@ list:
   just --list
 
 test:
-  cargo test skyproxy
+  cargo test skyproxy -- --nocapture
```
Comment (lynnliu030) [inline s3-proxy/justfile:14]: Why? change it back

File: store-server/operations/object_operations.py [hunk 2]
```diff
@@ -40,11 +41,61 @@
 from operations.utils.db import get_session, logger
 from typing import List
 from datetime import datetime
+from .policy.placement_policy import (
```
Comment (lynnliu030) [inline store-server/operations/object_operations.py:44]: remove `.policy` use `operations.policy`

File: store-server/operations/object_operations.py [hunk 3]
```diff
@@ -440,6 +491,8 @@ async def locate_object(
         )
     locators = (await db.scalars(stmt)).first()
 
+    # self.transfer_policy.get(req)
```
Comment (lynnliu030) [inline store-server/operations/object_operations.py:494]: Remove comments

File: store-server/operations/policy/placement_policy.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+class PlacementPolicy:
+    def __init__(self) -> None:
+        self.stat_graph = make_nx_graph()
+        pass
+
+    def place(self, req: StartUploadRequest) -> List[str]:
+        pass
+
+    def get_policy(self) -> str:
```
Comment (lynnliu030) [inline store-server/operations/policy/placement_policy.py:16]: Add comments to these functions (input, output), change `get_policy` to something like `def name(self) -> str`

File: store-server/operations/policy/placement_policy.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+
+class ReplicateAll(PlacementPolicy):
+    """
+    Replicate all objects to all regions
+    """
+
+    def place(self, req: StartUploadRequest) -> List[str]:
+        if req.op == "write":
+            return list(self.total_graph.nodes())
```
Comment (lynnliu030) [inline store-server/operations/policy/placement_policy.py:47]: there is no `total_graph` here, have you test all the policies?

File: store-server/operations/policy/placement_policy.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+        req: StartUploadRequest,
+        physical_bucket_locators: List[DBPhysicalBucketLocator],
+    ) -> List[str]:
+        upload_to_region_tags = [
+            locator.location_tag
+            for locator in physical_bucket_locators
+            if locator.is_primary or locator.need_warmup
+        ]
+        # check the type of physical_bucket_locators
+        # print("type of locators: ", type(physical_bucket_locators))
```
Comment (lynnliu030) [inline store-server/operations/policy/placement_policy.py:71]: Remove comments

File: store-server/operations/policy/placement_policy.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+        if req.op == "write":
+            return [req.client_from_region]
+        else:
+            return []
+
+    def get_policy(self) -> str:
+        return "write_local"
+
+
+put_policy = PlacementPolicy()
```
Comment (lynnliu030) [inline store-server/operations/policy/placement_policy.py:105]: What is this variable used for?

File: store-server/operations/policy/transfer_policy.py [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+
+    # def get_physical_object_locators(self, req: LocateObjectRequest) -> List[str]:
+    #     physical_object_locators = []
+
+    def get(
+        self, req: LocateObjectRequest, physical_locators: List[DBPhysicalObjectLocator]
+    ) -> DBPhysicalObjectLocator:
+        # Assume always direct transfer
+        # phyiscal_locators = self.get_physical_object_locators(req)
+        # place based on the physical locators
```
Comment (lynnliu030) [inline store-server/operations/policy/transfer_policy.py:21]: Remove these comments and the comments above; why don't we use the `def get_physical_object_locators` function to calculate the physical locators for the request? why do we need to pass in the list of physical locators here?

File: store-server/operations/policy/transfer_policy.py [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+        for locator in physical_locators:
+            if client_from_region == locator.location_tag:
+                return locator
+
+        # find the cheapest region to get from client_from_region
+        # can get the cost of self.stat_graph[client_from_region][region]['cost']
+        # can get the latency of self.stat_graph[client_from_region][region]['latency']
+        min_cost = sys.maxsize
+        chosen_locator = None
+        for locator in physical_locators:
```
Comment (lynnliu030) [inline store-server/operations/policy/transfer_policy.py:42]: `chosen_locator = min(physical_locators, key=lambda loc: self.stat_graph[client_from_region][loc.location_tag]["cost"])
`

File: store-server/operations/policy/transfer_policy.py [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+
+        print("type: ", physical_locators)
+
+        for locator in physical_locators:
+            if client_from_region == locator.location_tag:
+                return locator
+
+        # find the cheapest region to get from client_from_region
+        # can get the cost of self.stat_graph[client_from_region][region]['cost']
+        # can get the latency of self.stat_graph[client_from_region][region]['latency']
```
Comment (lynnliu030) [inline store-server/operations/policy/transfer_policy.py:65]: Remove redundant comments and print

File: store-server/operations/policy/transfer_policy.py [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+        for locator in physical_locators:
+            region = locator.location_tag
+            print("region: ", region)
+            if (
+                max_throughput
+                < self.stat_graph[client_from_region][region]["throughput"]
+            ):
+                max_throughput = self.stat_graph[client_from_region][region][
+                    "throughput"
+                ]
```
Comment (lynnliu030) [inline store-server/operations/policy/transfer_policy.py:77]: `chosen_locator = max(
    physical_locators, 
    key=lambda loc: self.stat_graph[client_from_region][loc.location_tag]["throughput"]
)
`

File: store-server/operations/policy/utils/test/test.py [hunk 12]
```diff
@@ -0,0 +1,12 @@
+from helpers import make_nx_graph
+import numpy as np
+
+G = make_nx_graph()
```
Comment (lynnliu030) [inline store-server/operations/policy/utils/test/test.py:4]: What is this file for? should remove it

File: store-server/operations/schemas/object_schemas.py [hunk 13]
```diff
@@ -192,7 +192,7 @@ class PatchUploadIsCompleted(BaseModel):
     etag: str
     last_modified: datetime
     version_id: Optional[str] = None
-    policy: Optional[str] = "push"
+    # policy: Optional[str] = "push"
```
Comment (lynnliu030) [inline store-server/operations/schemas/object_schemas.py:195]: Just delete instead of commenting it out

**Review by lynnliu030 — CHANGES_REQUESTED** [review]

File: store-server/operations/policy/placement_policy.py (hunk 7 above)
Comment (shaopu1225) [reply store-server/operations/policy/placement_policy.py:105]: This variable is to keep track of the current put_policy, the `update_policy` method will update this variable and the `get_policy` variable defined in `transfer_policy.py`.

File: store-server/experiment/trace/trace_output_new.csv [hunk 14]
```diff
@@ -0,0 +1,100 @@
+timestamp,op,issue_region,obj_key,size
+2023-12-01 18:08:50,write,aws:us-west-2,3f037af8,100722204
+2023-12-01 18:08:54,write,aws:sa-east-1,58320fcd,84610717
+2023-12-01 18:08:56,read,aws:ca-central-1,3f037af8,100722204
+2023-12-01 18:09:00,write,aws:eu-north-1,2a35b6a4,211151561
+2023-12-01 18:09:02,write,aws:eu-central-2,b7c3f4e4,439391083
```
Comment (lynnliu030) [inline store-server/experiment/trace/trace_output_new.csv:6]: This file is used for? maybe we can just leave one of these

File: store-server/operations/bucket_operations.py [hunk 15]
```diff
@@ -419,6 +420,8 @@ async def put_bucket_versioning(
 async def check_version_setting(
     request: HeadBucketRequest, db: Session = Depends(get_session)
 ) -> bool:
+    # print("start check time: ", datetime.now())
```
Comment (lynnliu030) [inline store-server/operations/bucket_operations.py:423]: Remove print (and all comments)

File: store-server/operations/policy/placement_policy.py [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+        pass
+
+    def place(self, req: StartUploadRequest) -> List[str]:
+        """
+        Args:
+            req: StartUploadRequest
+        Returns:
+            List[str]: single region to write to
+        """
+        single_store_region = "aws:us-west-1"
```
Comment (lynnliu030) [inline store-server/operations/policy/placement_policy.py:33]: Put a note here: hard coded for now; make this a variable init in `def __init__`

File: store-server/operations/policy/placement_policy.py [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+        req: StartUploadRequest,
+    ) -> List[str]:
+        """
+        Args:
+            req: StartUploadRequest
+        Returns:
+            List[str]: the regions to push to, including the primary region and the regions we want to push to
+        """
+        # assert all push regions in init_regions
+        push_regions = ["aws:us-west-1", "aws:us-east-1"]
```
Comment (lynnliu030) [inline store-server/operations/policy/placement_policy.py:83]: Same things for here

File: store-server/operations/policy/placement_policy.py [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+        Returns:
+            List[str]: the local region client is from
+        """
+        return [req.client_from_region]
+
+    def name(self) -> str:
+        return "write_local"
+
+
+def build_placement_policy_from_name(
```
Comment (lynnliu030) [inline store-server/operations/policy/placement_policy.py:129]: `get_policy()` would be enough

File: store-server/operations/policy/transfer_policy.py [hunk 19]
```diff
@@ -0,0 +1,119 @@
...
+
+class TransferPolicy:
+    def __init__(self) -> None:
+        self.stat_graph = make_nx_graph()
+        pass
+
+    def get(
+        self, req: LocateObjectRequest, physical_locators: List[DBPhysicalObjectLocator]
+    ) -> DBPhysicalObjectLocator:
+        print("TransferPolicy.get() called")
```
Comment (lynnliu030) [inline store-server/operations/policy/transfer_policy.py:14]: Remove

File: store-server/operations/policy/transfer_policy.py [hunk 20]
```diff
@@ -0,0 +1,119 @@
...
+        )
+        if locator:
+            return locator
+        pass
+
+    def name(self) -> str:
+        return "direct"
+
+
+def build_transfer_policy_from_name(name: str) -> TransferPolicy:
```
Comment (lynnliu030) [inline store-server/operations/policy/transfer_policy.py:111]: same name as above, `get_transfer_policy`

File: store-server/operations/schemas/bucket_schemas.py [hunk 21]
```diff
@@ -17,13 +18,16 @@
 class DBLogicalBucket(Base):
     __tablename__ = "logical_buckets"
 
-    id = Column(Integer, primary_key=True, autoincrement=True)
-    bucket = Column(String)
+    id = Column(Integer, primary_key=True, autoincrement=True, unique=True)
+    # id = Column(Integer, primary_key=True, autoincrement=True)
```
Comment (lynnliu030) [inline store-server/operations/schemas/bucket_schemas.py:22]: Can you make a note here on the Postgres and SQLite difference?

Comment (lynnliu030) [reply store-server/operations/schemas/bucket_schemas.py:22]: Or maybe is it better that you just copy the SQLite bucket_schemas file into another file?

File: store-server/operations/utils/db.py [hunk 22]
```diff
@@ -16,16 +23,127 @@
...
+        )
+        await conn.execute(f"CREATE DATABASE {db_name}")
+        print(f"Database {db_name} created successfully.")
+        await conn.execute(
+            f"ALTER DATABASE {db_name} SET DEFAULT_TRANSACTION_ISOLATION TO 'read committed';"
+        )
+        # await conn.execute(f"ALTER DATABASE {db_name} SET synchronous_commit TO off;")
+        # await conn.execute(f"ALTER DATABASE {db_name} SET wal_level TO minimal;")
+        # await conn.execute(f"ALTER DATABASE {db_name} SET fsync TO off;")
+        # await conn.execute(f"ALTER DATABASE {db_name} SET full_page_writes TO off;")
```
Comment (lynnliu030) [inline store-server/operations/utils/db.py:64]: Make a note here to say these are the optimizations can be applied in VM setup

**Review by lynnliu030 — CHANGES_REQUESTED** [review]

File: store-server/operations/schemas/bucket_schemas.py (hunk 21 above)
Comment (shaopu1225) [reply store-server/operations/schemas/bucket_schemas.py:22]: Sure,  I wll do that.

**Pull request merged**